engine/flink/util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/AggregateTransformer.scala

```diff
@@ -0,0 +1,127 @@
+              @OutputVariableName variableName: String)(implicit nodeId: NodeId): ContextTransformation = ContextTransformation
+    .definedBy(aggregator.toContextTransformation(variableName, aggregateBy))
+    .implementedBy(
+      FlinkCustomStreamTransformation((start: DataStream[NkContext], ctx: FlinkCustomNodeContext) => {
+        val windowMillis = Duration(length).toMillis
+        start
+          .map(new KeyWithValueMapper(ctx.lazyParameterHelper, keyBy, aggregateBy))
+          .keyBy(_.value._1)
+          .process(new AggregatorFunction[AnyRef](aggregator, windowMillis, ctx.nodeId))
+      }))
+}
+
+
+object AggregateTumblingTransformer extends CustomStreamTransformer {
+
+  @MethodToInvoke(returnType = classOf[AnyRef])
+  def execute(@ParamName("keyBy") keyBy: LazyParameter[String],
+              @ParamName("lengthInSe") length: String,
+              @ParamName("aggregator") aggregator: Aggregator,
+              @ParamName("aggregateBy") aggregateBy: LazyParameter[AnyRef],
+              @OutputVariableName variableName: String)(implicit nodeId: NodeId): ContextTransformation = ContextTransformation
+    .definedBy(aggregator.toContextTransformation(variableName, aggregateBy))
+    .implementedBy(
+      FlinkCustomStreamTransformation((start: DataStream[NkContext], ctx: FlinkCustomNodeContext) => {
+        val windowMillis = Duration(length).toMillis
+
+        start
+          .map(new KeyWithValueMapper(ctx.lazyParameterHelper, keyBy, aggregateBy))
+          .keyBy(_.value)
+          .window(TumblingProcessingTimeWindows.of(Time.milliseconds(windowMillis)))
+          .aggregate(aggregator
+            // this casting seems to be needed for Flink to infer TypeInformation correctly....
+            .asInstanceOf[org.apache.flink.api.common.functions.AggregateFunction[ValueWithContext[(String, AnyRef)], AnyRef, ValueWithContext[Any]]])
+      }))
+
+}
+
+//TODO: how to handle keyBy in more elegant way?
+class KeyWithValueMapper(val lazyParameterHelper: FlinkLazyParameterFunctionHelper, key: LazyParameter[String], value: LazyParameter[AnyRef])
+  extends RichMapFunction[NkContext, ValueWithContext[(String, AnyRef)]] with LazyParameterInterpreterFunction {
+
+  private lazy val keyInterpreter = lazyParameterInterpreter.syncInterpretationFunction(key)
+
+  private lazy val valueInterpreter = lazyParameterInterpreter.syncInterpretationFunction(value)
+
+  override def map(value: NkContext): ValueWithContext[(String, AnyRef)] = {
+    ValueWithContext((keyInterpreter(value), valueInterpreter(value)), value)
+  }
+}
+
+
+//TODO: figure out if it's more convenient/faster to just use SlidingWindow with 60s slide and appropriate trigger?
+class AggregatorFunction[T <: AnyRef](aggregator: Aggregator, lengthInMillis: Long, nodeId: String)
+  extends LatelyEvictableStateFunction[ValueWithContext[(String, AnyRef)], ValueWithContext[Any], TreeMap[Long, AnyRef]] {
+
+  private val minimalResolutionMs = 60000L
```

Comment:
Why you need this?

Reply:
to reduce the number of values we are storing. It's kind of window slide.